tools/setup_helpers/nccl.py

```diff
@@ -5,19 +5,23 @@
     gather_paths
 
 from .cuda import USE_CUDA, CUDA_HOME
+from .rocm import USE_ROCM, ROCM_HOME
 
-USE_NCCL = USE_CUDA and not check_negative_env_flag('USE_NCCL') and not IS_DARWIN and not IS_WINDOWS
+USE_NCCL = (USE_CUDA or USE_ROCM) and not check_negative_env_flag('USE_NCCL') and not IS_DARWIN and not IS_WINDOWS
```

Comment:
Please see #22464 . What `.rocm` has done is really inconsistent with the actual detection by CMake, and I don't think variables like `ROCM_HOME` is actually passed or used. We likely need to move this to CMakeLists.txt rather than doing handcrafted detection here.

Reply:
@xuhdev I'm assuming the nccl.py setup_helper file will need to stay, so it can detect if NCCL is actually installed for a CUDA/ROCm build. Is that correct? To do that, it will need to use USE_ROCM and ROCM_PATH (or the erstwhile ROCM_HOME) to locate the rccl header and rccl library. How do you suggest we do that now?

Reply:
I would suggest you to modify ` cmake/External/nccl.cmake`. `rocm.py` is not passing useful information down to CMake -- even if it stays, it's likely not gonna help.

Reply:
I was referring to `tools/setup_helpers/nccl.py`. It is used to locate the nccl/rccl library and the header nccl.h/rccl.h.

It is imported in setup.py as follows:
`from tools.setup_helpers.nccl import NCCL_SYSTEM_LIB, NCCL_INCLUDE_DIR`
**before** cmake is called, afaiu. 

`cmake/External/nccl.cmake` then uses the above variables:
```
  if (USE_SYSTEM_NCCL)
    # if we have explicit paths passed from setup.py, use those
    if (NCCL_INCLUDE_DIR)
```

In light of this, what's the future plan for the build flow? Is `tools/setup_helpers/nccl.py` going to continue to stay?

Reply:
#22480 will remove nccl.py.

It is unfortunate that PyTorch's build scripts are not in an organized way and we are on the way to clean these Python scripts up. The issue with rocm.py is that it seemingly accepts many options specified by users, but they are actually not effective or consistently effective. If your detection depends on rocm.py, what will happen is that users will have to "double specify" build options for ROCM, one for NCCL and one for the real ROCM.

Reply:
Okay, I'm going to wait "for the dust to settle down" on this build flow refactoring, otherwise I'll just run into more conflicts and have to rework. If possible, please let me know when the build flow refactoring that'd affect this PR is done. I'll keep an eye on #22480

Reply:
It looks like the cmake changes were finally part of #22930. I'm working on refactoring the changes in this PR in accordance with the cmake changes.